libtenzir/builtins/operators/deduplicate.cpp

```diff
@@ -122,33 +136,145 @@ class deduplicate_operator final : public crtp_operator<deduplicate_operator> {
   using match_store
     = std::unordered_map<data, match_type, match_hash, std::equal_to<>>;
 
-  auto make_indices(const type& schema) const -> std::vector<offset> {
+  auto make_projection(const table_slice& slice, diagnostic_handler& diag) const
+    -> std::optional<cached_projection> {
     std::vector<offset> indices{};
+    std::vector<std::string> missing_fields{};
+    const auto& schema = slice.schema();
     for (const auto& field : cfg_.fields) {
+      bool resolved = false;
       for (auto idx : schema.resolve(field)) {
         indices.emplace_back(std::move(idx));
+        resolved = true;
+      }
+      if (not resolved) {
+        TENZIR_ASSERT(not field.empty());
+        if (field.starts_with(':')) {
+          // We can't easily deal with missing type extractors, just erroring out.
+          diagnostic::error(
+            "failed to deduplicate due to unmatched type extractor")
+            .note("`{}` matched no fields in schema `{}`", field, schema.name())
+            .emit(diag);
+          return std::nullopt;
+        }
+        missing_fields.emplace_back(field);
       }
     }
-    std::sort(indices.begin(), indices.end());
-    indices.erase(std::unique(indices.begin(), indices.end()), indices.end());
-    return indices;
+    if (missing_fields.empty()) {
+      // Every field in `cfg_.fields` was found in `slice.schema()`.
+      // Clean up the indices and return.
+      std::ranges::sort(indices);
```

Comment:
It seems like we incorrectly depend on the order of the fields. Perhaps this was also the case before this PR, I didn't check that.

~~~
$ cat test.json
{"a": 1, "b": 2}
{"b": "reset"}
{"b": 2, "a": 1}

$ tenzir "from test.json | deduplicate a, b"
{
  "a": 1,
  "b": 2
}
{
  "b": "reset"
}
{
  "b": 2,
  "a": 1
}
~~~

(The middle event is needed to convince the JSON reader to use a different field ordering).

Reply:
Oh darn it, we do, and also did before

Reply:
I think this is now fixed. This PR is really starting to blow up in size, and in turn, so is `deduplicate` in complexity!